src/sentry/search/snuba/backend.py

```diff
@@ -0,0 +1,352 @@
+
+    if not d.tzinfo:
+        d = d.replace(tzinfo=pytz.utc)
+
+    return d
+
+
+def calculate_priority_cursor(data):
+    times_seen = sum(data['times_seen'])
+    last_seen = max(int(to_timestamp(snuba_str_to_datetime(d)) * 1000) for d in data['last_seen'])
+    return ((math.log(times_seen) * 600) + last_seen)
+
+
+def _datetime_cursor_calculator(field, fn):
+    def calculate(data):
+        datetimes = [snuba_str_to_datetime(d) for d in data[field]]
+        return fn(int(to_timestamp(d) * 1000) for d in datetimes)
+
+    return calculate
+
+
+sort_strategies = {
+    # sort_by -> Tuple[
+    #   String: expression to generate sort value (of type T, used below),
+    #   Function[T] -> int: function for converting a group's data to a cursor value),
+    # ]
+    'priority': (
+        '-priority', calculate_priority_cursor,
+    ),
+    'date': (
+        '-last_seen', _datetime_cursor_calculator('last_seen', max),
+    ),
+    'new': (
+        '-first_seen', _datetime_cursor_calculator('first_seen', min),
+    ),
+    'freq': (
+        '-times_seen', lambda data: sum(data['times_seen']),
+    ),
+}
+
+
+class SnubaConditionBuilder(object):
+    """\
+    Constructions a Snuba conditions list from a ``parameters`` mapping.
+
+    ``Condition`` objects are registered by their parameter name and used to
+    construct the Snuba condition list if they are present in the ``parameters``
+    mapping.
+    """
+
+    def __init__(self, conditions):
+        self.conditions = conditions
+
+    def build(self, parameters):
+        result = []
+        for name, condition in self.conditions.items():
+            if name in parameters:
+                result.append(condition.apply(name, parameters))
+        return result
+
+
+class Condition(object):
+    """\
+    Adds a single condition to a Snuba conditions list. Used with
+    ``SnubaConditionBuilder``.
+    """
+
+    def apply(self, name, parameters):
+        raise NotImplementedError
+
+
+class CallbackCondition(Condition):
+    def __init__(self, callback):
+        self.callback = callback
+
+    def apply(self, name, parameters):
+        return self.callback(parameters[name])
+
+
+class ScalarCondition(Condition):
+    """\
+    Adds a scalar filter (less than or greater than are supported) to a Snuba
+    condition list. Whether or not the filter is inclusive is defined by the
+    '{parameter_name}_inclusive' parameter.
+    """
+
+    def __init__(self, field, operator, default_inclusivity=True):
+        assert operator in ['<', '>']
+        self.field = field
+        self.operator = operator
+        self.default_inclusivity = default_inclusivity
+
+    def apply(self, name, parameters):
+        inclusive = parameters.get(
+            '{}_inclusive'.format(name),
+            self.default_inclusivity,
+        )
+
+        arg = parameters[name]
+        if isinstance(arg, datetime):
+            if not arg.tzinfo:
+                arg = arg.replace(tzinfo=pytz.utc)
```

Comment:
Can't speak for this specifically, but because we use the utils in `sentry/utils/dates.py` we need to pass in TZ-aware datetimes. and confusingly `datetime.urcnow()` returns a naive datetime, not a UTC datetime.

Reply:
Yeah, it was `to_datetime` failing when it tried to do math on naive and non-naive. I assumed the only right thing to do with a naive datetime was to treat it like UTC?

Reply:
Makes sense. I was more specifically wondering where it came from (it seems like we shouldn't be passing naive datetimes into this function in the first place and should probably fix that upstream.)